File: platforms/android/app/src/main/kotlin/hk/noq/cordova/plugin/RoomQPlugin.kt
package hk.noq.cordova.plugin;

// Cordova-required packages
import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.PluginResult;
import org.apache.cordova.PermissionHelper;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

// Android packages
import android.Manifest;
import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.content.pm.PackageManager;

import android.content.Context;

// RoomQ SDK
import hk.noq.roomq.*
import hk.noq.roomq.model.*

class RoomQPlugin : CordovaPlugin() {
  lateinit var callbackContext: CallbackContext;
  lateinit var roomQ: RoomQ;
  lateinit var waitingRoomIntent: Intent;

  companion object {
    protected val LOG_TAG = "RoomQ Plugin";
    protected val INTERNET = Manifest.permission.INTERNET;
    protected val REQUEST_WAITING_ROOM = 0;
  }

  override fun execute(action: String, data: JSONArray, callbackContext: CallbackContext): Boolean {
    this.callbackContext = callbackContext;

    var result = true;
    try {
      when (action) {
        "initRoomQ" -> {
          var clientID = data.getString(0);
          this.initRoomQ(clientID);
        }
        "enqueue" -> this.enqueue();
        "getExpiryTime" -> this.getExpiryTime();
        "extendSession" -> {
          var minutes = data.getInt(0);
          this.extendSession(minutes);
        }
        "deleteSession" -> this.deleteSession();
        "setClientToken" -> {
          var token = data.getString(0);
          this.setClientToken(token);
        }
        else -> {
          handleError("Invalid action");
          result = false;
        }
      }
    } catch (e: Exception) {
      handleException(e);
      result = false;
    }
    return result;
  }

  override fun onRestoreStateForActivityResult(state: Bundle, callbackContext: CallbackContext) {
    this.callbackContext = callbackContext;
  }

  private fun initRoomQ(clientID: String) {
    roomQ = RoomQ(clientID);
    callbackContext.success();
  }

  private fun enqueue() {
    try {
      roomQ.enqueue(getContext()) { result ->
        when (result) {
          is EnqueueResult.Wait -> {
            waitingRoomIntent = result.intent;
            openWaitingRoomActivity();
            callbackContext.success();
          }
          is EnqueueResult.Enter -> {
            Log.d(LOG_TAG, "enter normal flow");
            callbackContext.success();
          }
          is EnqueueResult.Error -> {
            Log.d(LOG_TAG, "enter normal flow");
            callbackContext.success()
          }
        }
      }
    } catch (error: Error) {

    }

  }

  private fun getExpiryTime() {
    roomQ.getExpiryTime(getContext()) { response ->
      when (response) {
        is GetExpiryTimeResult.Success -> {
          Log.d(LOG_TAG, "case 1: if the session is still valid, it will return the deadline via response.deadline");
          callbackContext.success()
        }
        GetExpiryTimeResult.Expired -> {
          Log.d(LOG_TAG, "case 2: the session is already expired before extension, to proceed to waiting room, please call enqueue again");
          callbackContext.success()
        }
        GetExpiryTimeResult.NoToken -> {
          Log.d(LOG_TAG, "case 3: no token obtained before");
          callbackContext.success()
        }
        is GetExpiryTimeResult.Error -> {
          Log.d(LOG_TAG, "case 4: Error occur in SDK");
          handleError("Error occur in SDK")
        }
      }
    }
  }

  private fun extendSession(minutes: Int) {
    roomQ.extendSession(getContext(), minutes) { response ->
      when (response) {
        ExtendSessionResponse.EXTENDED -> {
          Log.d(LOG_TAG, "case 1: the session is extended successfully");
          callbackContext.success()
        }
        ExtendSessionResponse.EXPIRED -> {
          Log.d(LOG_TAG, "case 2: the session is already expired before extension, to proceed to waiting room, please call enqueue again");
          callbackContext.success()
        }
        ExtendSessionResponse.SERVER_ERROR -> {
          Log.d(LOG_TAG, "case 3: Error occur in SDK");
          handleError("Error occur in SDK")
        }
      }
    }
  }

  private fun deleteSession() {
    roomQ.deleteSession(getContext()) { response ->
      when (response) {
        DeleteSessionResponse.SUCCESS -> {
          Log.d(LOG_TAG, "case 1: the session is deleted successfully");
          callbackContext.success()
        }
        DeleteSessionResponse.SERVER_ERROR -> {
          Log.d(LOG_TAG, "case 2: Error occur in SDK");
          handleError("Error occur in SDK")
        }
      }
    }
  }

  private fun setClientToken(token: String) {
    roomQ.setToken(this.getContext(), token)
  }

  private fun getContext(): Context {
    return cordova.getActivity().getApplicationContext()
  }

  private fun openWaitingRoomActivity() {
    cordova.setActivityResultCallback(this);
    cordova.startActivityForResult(this, waitingRoomIntent, REQUEST_WAITING_ROOM);
  }

  override fun onActivityResult(requestCode: Int, resultCode: Int, intent: Intent?) {
    if (requestCode == REQUEST_WAITING_ROOM) {
      if (resultCode == Activity.RESULT_OK && intent != null) {
        val token = intent.getStringExtra(RoomQ.TOKEN);
        Log.d(LOG_TAG, "enter normal flow");
        Log.d(LOG_TAG, token);
      }
    }
  }

  private fun handleError(errorMsg: String) {
    try {
      Log.e(LOG_TAG, errorMsg);
      callbackContext.error(errorMsg);
    } catch (e: Exception) {
      Log.e(LOG_TAG, e.toString());
    }
  }

  private fun handleException(exception: Exception) {
    handleError(exception.toString());
  }
}

File: platforms/ios/RoomQPluginTest/Plugins/cordova-plugin-roomq/RoomQPlugin.swift
import Foundation
import roomq

@objc(RoomQPlugin) class RoomQPlugin: CDVPlugin {
  var roomq: RoomQ!

  @objc(initRoomQ:)
  func initRoomQ(command: CDVInvokedUrlCommand) {
    NSLog("RoomQPlugin#initRoomQ()")
    if (roomq != nil) {
      self.emit(command.callbackId, result: CDVPluginResult(status: CDVCommandStatus_OK, messageAs: "The plugin succeeded"))
      return
    }
    
    let cliendID = command.arguments[0] as! String

    if (cliendID == "") {
      self.emit(command.callbackId, result: CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: "You must provide valid client ID"))
    }

    self.roomq = RoomQ(clientID: cliendID)
    self.emit(command.callbackId, result: CDVPluginResult(status: CDVCommandStatus_OK, messageAs: "The plugin succeeded"))
  }

  @objc(enqueue:)
  func enqueue(command: CDVInvokedUrlCommand) {
    NSLog("RoomQPlugin#enqueue()")

    if (roomq == nil) {
      self.emit(command.callbackId, result: CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: "You must call initRoomQ before this function."))
      return
    }
    
    self.roomq.enqueue { [weak self] result in
    switch result {
      case let .wait(vc, token):
        self?.showWaitingRoom(vc)
      case let .enter(source, token):
        switch source {
          case .directly, .error:
            NSLog("enter normal flow")
          default:
            NSLog("enter normal flow")
        }
      case let .error(error):
        NSLog(error.localizedDescription)
    }
    self.emit(command.callbackId, result: CDVPluginResult(status: CDVCommandStatus_OK, messageAs: "The plugin succeeded"))
    }
  }
    
  @objc(setClientToken:)
  func setClientToken(command: CDVInvokedUrlCommand) {
    NSLog("RoomQPlugin#setClientToken()")
    if (roomq == nil) {
      self.emit(command.callbackId, result: CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: "You must call initRoomQ before this function."))
      return
    }
    
    let token = command.arguments[0] as! String
    
    self.roomq.set(token: token)
    self.emit(command.callbackId, result: CDVPluginResult(status: CDVCommandStatus_OK, messageAs: "Client token is set"))
  }
    
  @objc(getExpiryTime:)
  func getExpiryTime(command: CDVInvokedUrlCommand) {
    NSLog("RoomQPlugin#getExpiryTime()")
    if (roomq == nil) {
      self.emit(command.callbackId, result: CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: "You must call initRoomQ before this function."))
      return
    }
    
    self.roomq.getExpiryTime { [weak self] result in
      switch result {
        case .success(let deadline):
          NSLog("case 1: if the session is still valid, it will return the deadline")
        case .expired:
          NSLog("case 2: the session is already expired")
        case .noToken:
          NSLog("case 3: no token obtained before")
        case .error:
          NSLog("case 4: Error occur in SDK")
      }
    }
  }
    
  @objc(extendSession:)
  func extendSession(command: CDVInvokedUrlCommand) {
    NSLog("RoomQPlugin#extendSession()")
    if (roomq == nil) {
      self.emit(command.callbackId, result: CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: "You must call initRoomQ before this function."))
      return
    }
    
    let minutes = command.arguments[0] as Int
    
    self.roomq.extendSession(minutes: minutes) { [weak self] result in
      switch result {
        case .extended:
          NSLog("case 1: the session is extended successfully")
        case .expired:
          NSLog("case 2: the session is already expired before extension, to proceed to waiting room, please call enqueue again")
        case .invalidExtension:
          NSLog("case 3: the session is not extended because the specified extension duration is invalid")
        case .serverError:
          NSLog("case 4: Error occur in SDK")
      }
    }
  }
    
  @objc(deleteSession:)
  func deleteSession(command: CDVInvokedUrlCommand) {
    NSLog("RoomQPlugin#deleteSession()")
    if (roomq == nil) {
      self.emit(command.callbackId, result: CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: "You must call initRoomQ before this function."))
      return
    }
    
    self.roomq.deleteSession(callback: { [weak self] result in
      switch result {
        case .success:
          NSLog("case 1: the session is deleted successfully")
        case .serverError:
          NSLog("case 2: Error occur in SDK")
      }
    })
  }

  func showWaitingRoom(_ vc: UIViewController) {
    viewController.present(vc, animated: true, completion: nil)
  }

  fileprivate func emit(_ callbackId: String, result: CDVPluginResult) {
    self.commandDelegate!.send(result, callbackId: callbackId)
  }
}